File: typescript/core/test/stream.test.ts
/**
 * Unit tests for streaming turn execution (`SmoothAgent.runStream`).
 *
 * Proves the streaming loop mirrors the C# `RunStreamingAsync` behaviour: text
 * deltas surface as multiple `text` events, tool calls round-trip through the same
 * dispatch path (clearance/human-gate/JSON parsing), arguments split across chunks
 * still assemble correctly, and the terminal `done` event carries a response
 * equivalent to non-streaming `run()`.
 */

import { describe, expect, it } from 'vitest';
import { AgentOptions, SmoothAgent, StreamEvent, Tool } from '../src/agent.js';
import { MockLlmProvider } from '../src/llmProvider.js';

async function collect(gen: AsyncGenerator<StreamEvent>): Promise<StreamEvent[]> {
    const events: StreamEvent[] = [];
    for await (const e of gen) events.push(e);
    return events;
}

describe('SmoothAgent.runStream', () => {
    it('streams a multi-chunk text reply then exactly one done', async () => {
        const mock = new MockLlmProvider();
        mock.pushText('hello there friend, how are you', { prompt_tokens: 10, completion_tokens: 7 });

        const agent = new SmoothAgent(mock, {} satisfies AgentOptions);
        const events = await collect(agent.runStream('hi'));

        const textEvents = events.filter((e) => e.type === 'text');
        expect(textEvents.length).toBeGreaterThanOrEqual(2);
        const joined = textEvents.map((e) => (e as { text: string }).text).join('');
        expect(joined).toBe('hello there friend, how are you');

        const doneEvents = events.filter((e) => e.type === 'done');
        expect(doneEvents).toHaveLength(1);
        expect(events[events.length - 1].type).toBe('done');
        const done = doneEvents[0] as { response: { text: string } };
        expect(done.response.text).toBe('hello there friend, how are you');
    });

    it('round-trips a tool call: tool_call event, tool runs, tool_result, final answer', async () => {
        let ran = '';
        const echo: Tool = {
            name: 'echo',
            description: 'Echoes input',
            parameters: { type: 'object', properties: { text: { type: 'string' } }, required: ['text'] },
            execute: async (args) => {
                ran = String(args.text ?? '');
                return `echoed:${ran}`;
            },
        };
        const mock = new MockLlmProvider();
        mock.pushToolCall('call-1', 'echo', '{"text":"ping"}', { prompt_tokens: 5, completion_tokens: 3 });
        mock.pushText('all done', { prompt_tokens: 8, completion_tokens: 2 });

        const agent = new SmoothAgent(mock, { tools: [echo] } satisfies AgentOptions);
        const events = await collect(agent.runStream('use echo'));

        const toolCall = events.find((e) => e.type === 'tool_call') as { name: string; arguments: string } | undefined;
        expect(toolCall?.name).toBe('echo');
        expect(JSON.parse(toolCall!.arguments)).toEqual({ text: 'ping' });

        // The tool actually ran with the assembled args.
        expect(ran).toBe('ping');

        const toolResult = events.find((e) => e.type === 'tool_result') as { name: string; result: string } | undefined;
        expect(toolResult?.name).toBe('echo');
        expect(toolResult?.result).toBe('echoed:ping');

        const done = events.find((e) => e.type === 'done') as { response: { text: string; iterations: number; toolCalls: number } };
        expect(done.response.text).toBe('all done');
        expect(done.response.iterations).toBe(2);
        expect(done.response.toolCalls).toBe(1);
    });

    it('assembles tool-call arguments split across chunks before dispatch', async () => {
        let received: Record<string, unknown> | undefined;
        const tool: Tool = {
            name: 'save',
            description: 'Saves',
            parameters: { type: 'object', properties: { key: { type: 'string' }, value: { type: 'string' } } },
            execute: async (args) => {
                received = args;
                return 'saved';
            },
        };
        const mock = new MockLlmProvider();
        // The mock splits these arguments across two chunks; the agent must reassemble them.
        mock.pushToolCall('call-1', 'save', '{"key":"alpha","value":"beta-gamma-delta"}');
        mock.pushText('ok');

        const agent = new SmoothAgent(mock, { tools: [tool] } satisfies AgentOptions);
        await collect(agent.runStream('save it'));

        expect(received).toEqual({ key: 'alpha', value: 'beta-gamma-delta' });
    });

    it('done carries a response equivalent to run() for the same script', async () => {
        const script = (): MockLlmProvider => new MockLlmProvider().pushText('the answer is 42', { prompt_tokens: 12, completion_tokens: 6 });
        const opts: AgentOptions = { model: 'claude-haiku-4-5' };

        const runResult = await new SmoothAgent(script(), opts).run('q');
        const events = await collect(new SmoothAgent(script(), opts).runStream('q'));
        const done = events.find((e) => e.type === 'done') as { response: typeof runResult };

        expect(done.response.text).toBe(runResult.text);
        expect(done.response.iterations).toBe(runResult.iterations);
        expect(done.response.toolCalls).toBe(runResult.toolCalls);
        expect(done.response.usage).toEqual(runResult.usage);
        expect(done.response.costUsd).toBeCloseTo(runResult.costUsd, 10);
        expect(runResult.costUsd).toBeGreaterThan(0);
    });

    it('throws if the client cannot stream', async () => {
        const nonStreaming = {
            chat: { completions: { create: async () => ({ choices: [{ message: { content: 'x' } }] }) } },
        };
        const agent = new SmoothAgent(nonStreaming, {});
        await expect(collect(agent.runStream('hi'))).rejects.toThrow(/streaming-capable/);
    });
});

File: go/core/stream_test.go
package core

import (
	"context"
	"encoding/json"
	"math"
	"testing"
)

// drain consumes a stream to completion, returning its events and the terminal error.
func drain(s *Stream) ([]StreamEvent, error) {
	var events []StreamEvent
	for e := range s.Events() {
		events = append(events, e)
	}
	return events, s.Err()
}

func TestRunStreamTextInMultipleChunks(t *testing.T) {
	mock := NewMockLlmProvider().PushResponse(WithUsage(TextResponse("hello there friend, how are you"), 10, 7))
	agent := NewSmoothAgent(mock, AgentOptions{})

	stream, err := agent.RunStream(context.Background(), "hi", nil)
	if err != nil {
		t.Fatal(err)
	}
	events, err := drain(stream)
	if err != nil {
		t.Fatal(err)
	}

	var text string
	textEvents := 0
	doneEvents := 0
	for _, e := range events {
		switch e.Kind {
		case StreamText:
			text += e.Text
			textEvents++
		case StreamDone:
			doneEvents++
		}
	}
	if textEvents < 2 {
		t.Fatalf("want >=2 text events, got %d", textEvents)
	}
	if text != "hello there friend, how are you" {
		t.Fatalf("concatenated text mismatch: %q", text)
	}
	if doneEvents != 1 {
		t.Fatalf("want exactly 1 done event, got %d", doneEvents)
	}
	if events[len(events)-1].Kind != StreamDone {
		t.Fatalf("done must be last")
	}
	if events[len(events)-1].Response.Text != "hello there friend, how are you" {
		t.Fatalf("done.Response.Text mismatch: %q", events[len(events)-1].Response.Text)
	}
}

func TestRunStreamToolRoundTrip(t *testing.T) {
	var ran string
	echo := FuncTool{
		ToolName: "echo",
		Desc:     "Echoes input",
		Params:   map[string]any{"type": "object", "properties": map[string]any{"text": map[string]any{"type": "string"}}},
		Fn: func(_ context.Context, args map[string]any) (string, error) {
			ran, _ = args["text"].(string)
			return "echoed:" + ran, nil
		},
	}
	mock := NewMockLlmProvider()
	mock.PushResponse(WithUsage(ToolCallResponse("call-1", "echo", `{"text":"ping"}`), 5, 3)).
		PushResponse(WithUsage(TextResponse("all done"), 8, 2))
	agent := NewSmoothAgent(mock, AgentOptions{Tools: []Tool{echo}})

	stream, err := agent.RunStream(context.Background(), "use echo", nil)
	if err != nil {
		t.Fatal(err)
	}
	events, err := drain(stream)
	if err != nil {
		t.Fatal(err)
	}

	var toolCall, toolResult, done *StreamEvent
	for i := range events {
		switch events[i].Kind {
		case StreamToolCall:
			toolCall = &events[i]
		case StreamToolResult:
			toolResult = &events[i]
		case StreamDone:
			done = &events[i]
		}
	}
	if toolCall == nil || toolCall.Name != "echo" {
		t.Fatalf("missing/incorrect tool_call event: %+v", toolCall)
	}
	var parsed map[string]any
	if err := json.Unmarshal([]byte(toolCall.Arguments), &parsed); err != nil {
		t.Fatalf("tool_call arguments not valid JSON: %v", err)
	}
	if parsed["text"] != "ping" {
		t.Fatalf("tool_call arguments mismatch: %+v", parsed)
	}
	if ran != "ping" {
		t.Fatalf("tool did not run with assembled args; ran=%q", ran)
	}
	if toolResult == nil || toolResult.Result != "echoed:ping" {
		t.Fatalf("missing/incorrect tool_result event: %+v", toolResult)
	}
	if done == nil || done.Response.Text != "all done" || done.Response.Iterations != 2 || done.Response.ToolCalls != 1 {
		t.Fatalf("done response mismatch: %+v", done)
	}
}

func TestRunStreamArgumentsSplitAcrossChunks(t *testing.T) {
	var received map[string]any
	save := FuncTool{
		ToolName: "save",
		Desc:     "Saves",
		Params:   map[string]any{"type": "object"},
		Fn: func(_ context.Context, args map[string]any) (string, error) {
			received = args
			return "saved", nil
		},
	}
	mock := NewMockLlmProvider()
	// The mock splits these arguments across two chunks; the agent must reassemble them.
	mock.PushToolCall("call-1", "save", `{"key":"alpha","value":"beta-gamma-delta"}`).PushText("ok")
	agent := NewSmoothAgent(mock, AgentOptions{Tools: []Tool{save}})

	stream, err := agent.RunStream(context.Background(), "save it", nil)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := drain(stream); err != nil {
		t.Fatal(err)
	}

	if received["key"] != "alpha" || received["value"] != "beta-gamma-delta" {
		t.Fatalf("arguments not reassembled: %+v", received)
	}
}

func TestRunStreamDoneMatchesRun(t *testing.T) {
	script := func() *MockLlmProvider {
		return NewMockLlmProvider().PushResponse(WithUsage(TextResponse("the answer is 42"), 12, 6))
	}
	opts := AgentOptions{Model: "claude-haiku-4-5"}

	runResult, err := NewSmoothAgent(script(), opts).Run(context.Background(), "q", nil)
	if err != nil {
		t.Fatal(err)
	}

	stream, err := NewSmoothAgent(script(), opts).RunStream(context.Background(), "q", nil)
	if err != nil {
		t.Fatal(err)
	}
	events, err := drain(stream)
	if err != nil {
		t.Fatal(err)
	}
	var done *StreamEvent
	for i := range events {
		if events[i].Kind == StreamDone {
			done = &events[i]
		}
	}
	if done == nil {
		t.Fatal("no done event")
	}
	r := done.Response
	if r.Text != runResult.Text || r.Iterations != runResult.Iterations || r.ToolCalls != runResult.ToolCalls {
		t.Fatalf("done response != run: stream=%+v run=%+v", r, runResult)
	}
	if r.Usage != runResult.Usage {
		t.Fatalf("usage mismatch: stream=%+v run=%+v", r.Usage, runResult.Usage)
	}
	if math.Abs(r.CostUSD-runResult.CostUSD) > 1e-9 {
		t.Fatalf("cost mismatch: stream=%v run=%v", r.CostUSD, runResult.CostUSD)
	}
	if runResult.CostUSD <= 0 {
		t.Fatalf("expected non-zero cost, got %v", runResult.CostUSD)
	}
}

func TestRunStreamRequiresStreamingClient(t *testing.T) {
	// A ChatClient that does not implement StreamingChatClient.
	agent := NewSmoothAgent(nonStreamingClient{}, AgentOptions{})
	if _, err := agent.RunStream(context.Background(), "hi", nil); err == nil {
		t.Fatal("expected error for non-streaming client")
	}
}

type nonStreamingClient struct{}

func (nonStreamingClient) Chat(_ context.Context, _ ChatRequest) (ChatResponse, error) {
	return TextResponse("x"), nil
}
